pkg/deploy/graph/analysis/dc.go

```diff
@@ -0,0 +1,87 @@
+package analysis
+
+import (
+	"fmt"
+
+	"github.com/gonum/graph"
+
+	osgraph "github.com/openshift/origin/pkg/api/graph"
+	buildedges "github.com/openshift/origin/pkg/build/graph"
+	buildgraph "github.com/openshift/origin/pkg/build/graph/nodes"
+	deployedges "github.com/openshift/origin/pkg/deploy/graph"
+	deploygraph "github.com/openshift/origin/pkg/deploy/graph/nodes"
+	imageedges "github.com/openshift/origin/pkg/image/graph"
+	imagegraph "github.com/openshift/origin/pkg/image/graph/nodes"
+)
+
+const (
+	ImageStreamTagNotAvailableWarning = "ImageStreamTagNotAvailable"
+	MissingImageStreamWarning         = "MissingImageStream"
+	MissingImageStreamTagWarning      = "MissingImageStreamTag"
+)
+
+// FindDeploymentConfigTriggerErrors checks for possible failures in deployment config
+// image change triggers.
+//
+// Precedence of failures:
+// 1. The image stream of the tag of interest does not exist.
+// 2. The image stream tag does not exist but a build config points to it.
+// 3. The image stream tag does not exist.
+func FindDeploymentConfigTriggerErrors(g osgraph.Graph) []osgraph.Marker {
+	markers := []osgraph.Marker{}
+
+dc:
+	for _, uncastDcNode := range g.NodesByKind(deploygraph.DeploymentConfigNodeKind) {
+		for _, uncastIstNode := range g.PredecessorNodesByEdgeKind(uncastDcNode, deployedges.TriggersDeploymentEdgeKind) {
+			if istNode := uncastIstNode.(*imagegraph.ImageStreamTagNode); !istNode.Found() {
+				dcNode := uncastDcNode.(*deploygraph.DeploymentConfigNode)
+
+				// 1. Image stream for tag of interest does not exist
+				for _, uncastIsNode := range g.SuccessorNodesByEdgeKind(uncastIstNode, imageedges.ReferencedImageStreamGraphEdgeKind) {
+					isNode := uncastIsNode.(*imagegraph.ImageStreamNode)
+					if isNode.Found() {
+						break
+					}
+					markers = append(markers, osgraph.Marker{
+						Node:         uncastDcNode,
+						RelatedNodes: []graph.Node{uncastIstNode, uncastIsNode},
+
+						Severity: osgraph.WarningSeverity,
+						Key:      MissingImageStreamWarning,
+						Message: fmt.Sprintf("The image trigger for %s will have no effect because %s does not exist.",
+							dcNode.ResourceString(), isNode.ResourceString()),
+					})
+					continue dc
+				}
+
+				// 2. Build config points to image stream tag of interest
```

Comment:
> What action should a user take in response to this warning?

EIther none, or in case the build fails to push, check what's wrong with it. (We may already inform about it too).